Implement Sandbox notifications processor and publisher

Future-Outlier opened the pull request:
# TL;DR
Implement notifications processor and publisher for Sandbox.


## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Add new **SandboxProcessor** and **SandboxPublisher** that implement the Processor interface.

I add config in yaml file like this
```
notifications:
  type: sandbox
  emailer:
    emailServerConfig:
      apiKeyEnvVar: "SENDGRID_API_KEY"
    subject: "Notice: Execution \"{{ name }}\" has {{ phase }} in \"{{ domain }}\"."
    sender: "example@gmail.com"  # Consider using a generic email instead of personal
    body: >
      Execution \"{{ name }}\" has {{ phase }} in \"{{ domain }}\". View details at
      <a href=\http://example.com/projects/{{ project }}/domains/{{ domain }}/executions/{{ name }}>
      http://example.com/projects/{{ project }}/domains/{{ domain }}/executions/{{ name }}</a>. {{ error }}
```
screenshot
![image](https://github.com/flyteorg/flyteadmin/assets/76461262/00c0bc6c-38b0-4c58-aa9c-000eeecc6484)

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

pingsutw commented on pkg/async/notifications/implementations/sandbox_processor.go:
```diff
@@ -0,0 +1,59 @@
+	return nil
+}
+
+func NewSandboxProcessor(emailer interfaces.Emailer) interfaces.Processor {
```
Could we add a test for the sandbox processor that uses the default noop emailer?

Future-Outlier replied:
Yes, I will try it

pingsutw commented on pkg/async/notifications/factory.go:
```diff
@@ -66,6 +66,8 @@ func GetEmailer(config runtimeInterfaces.NotificationsConfig, scope promutils.Sc
 			scope,
 			sesClient,
 		)
+	case common.Sandbox:
+		return implementations.NewSendGridEmailer(config, scope)
```
IIUC, we'll create SendGridEmailer at line 50, so we don't need these two lines here.

Future-Outlier replied:
Yes, but you need to turn your config to 
```
emailServerConfig:
      serviceName: sendgrid
      apiKeyEnvVar: "SENDGRID_API_KEY"
```
with one more key in serviceName

Future-Outlier commented:
@pingsutw  I've updated my test, please check it.
Also to use the mock emailer int test function, I've added a "return nil" in my sandbox_processor.go.
Thanks a lot.

Future-Outlier commented:
I am not pretty sure whether use noop emailer or mock emailer, please contact me if it is needed to change the code.
Thanks a lot.

Future-Outlier commented:
BTW, I just copy the master branch's go.mod and go.sum to my branch currently and run "go mod tidy"
I found that it is not needed to update the go.mod and go.sum.
Thanks a lot

pingsutw commented on pkg/async/notifications/implementations/sandbox_publisher_test.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+func TestSandboxPublisher_Publish(t *testing.T) {
+	publisher := NewSandboxPublisher()
```
Can we use NewNotificationsPublisher to create SandboxPublisher here?

Future-Outlier replied:
@pingsutw 
I've tried it, but the import cycle is not allowed in go test.

```
imports github.com/flyteorg/flyteadmin/pkg/async/notifications: import cycle not allowed in test
```
I've wrote code like this
```
var scope = promutils.NewScope("test_sandbox_processor")
notificationsConfig := runtimeInterfaces.NotificationsConfig{
Type: "sandbox",
}
testSandboxProcessor := notifications.NewNotificationsProcessor(notificationsConfig, scope)
```

And the dependencies I will list at here
```
package github.com/flyteorg/flyteadmin/pkg/async/notifications/implementations
imports github.com/flyteorg/flyteadmin/pkg/async/notifications: import cycle not allowed in test
```

If test the NewNotificationsPublisher  is necessary, I can try it in another package, but I am not sure whether this is what you want, thanks a lot

pingsutw replied:
I see, ok, that's fine

pingsutw commented on pkg/async/notifications/implementations/sandbox_processor_test.go:
```diff
@@ -0,0 +1,39 @@
+
+func TestSandboxProcessor_StartProcessing(t *testing.T) {
+
+	testSandboxProcessor := NewSandboxProcessor(&mockSandboxEmailer)
```
Same, Can we use NewNotificationsProcessor to create SandboxProcessor here?

pingsutw commented on pkg/async/notifications/factory_test.go:
```diff
@@ -23,3 +31,13 @@ func TestGetEmailer(t *testing.T) {
 }
+
+func TestNewNotificationsProcessor(t *testing.T) {
+	testSandboxProcessor := NewNotificationsProcessor(notificationsConfig, scope)
```
Nit: could you call StartProcessing and StopProcessing here?

Future-Outlier replied:
No problem

pingsutw reviewed (approved):
LGTM, Thanks!

Future-Outlier commented:
> LGTM, Thanks!

Thanks a lot for guiding me ^^

katrogan commented on pkg/async/notifications/implementations/sandbox_publisher.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+type SandboxPublisher struct{}
+
+var msgChan = make(chan []byte)
```
can we instantiate this in the `NewSandboxPublisher()` constructor and make it a member variable? Should make things easier to unit test too!

katrogan replied:
Alternatively we can pass it to NewSandboxPublisher() and NewSandboxProcessor()

Future-Outlier replied:
I think I can implement the alternative solution.
But how to pass the msgChan to first solution ?
Thanks a lot.

katrogan replied:
Sure, you could do 

```
// Receiver only channel
func NewSandboxProcessor(msgChan <-chan []byte)  

// Sender only channel
and NewSandboxPublisher(msgChan chan<- []byte)
```

And continue to initialize e chan it outside of the processor and publisher

Future-Outlier replied:
Can I clarify what "continue to initialize msgChan" means?
I thought that I just have to initialize the msgChan 1 time only.

katrogan commented on pkg/async/notifications/implementations/sandbox_processor.go:
```diff
@@ -0,0 +1,62 @@
+
+			err = p.email.SendEmail(context.Background(), emailMessage)
+			if err != nil {
+				logger.Errorf(context.Background(), "error with sendemail message [%v] ", err)
```
can we log the failed email message here too?

Future-Outlier replied:
Yes I will

katrogan commented on pkg/async/notifications/implementations/sandbox_processor_test.go:
```diff
@@ -0,0 +1,39 @@
+var mockSandboxEmailer mocks.MockEmailer
+
+func TestSandboxProcessor_UnmarshalMessage(t *testing.T) {
+	var emailMessage admin.EmailMessage
```
What's the purpose of this test? Are we just verifying we can unmarshal a proto? Considering that test code already initializes this msg var I'm not sure how valuable this unit test is.

Future-Outlier replied:
Yes, you are right, I can consider deleting it.

katrogan reviewed (approved):
Looks great, I think the code cov annotations have suggestions for some additional test cases we could use to bump the code coverage on this PR and so that we can test more scenarios

Future-Outlier commented:
> Looks great. This is failing the code coverage check, but I think the code cov annotations have suggestions for some additional test cases we could use to bump the code coverage on this PR and so that we can test more scenarios

Thanks for the suggestions, I've learned a lot!

katrogan reviewed (approved):
Thank you for adding all the unit tests!
